[strict-boolean-expressions] False positive with union types having only one falsey value

- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```javascript
module.exports = {
    "root": true,
    "parser": "@typescript-eslint/parser",
    "parserOptions": {
        "tsconfigRootDir": __dirname,
        "project": ["./tsconfig.json"]
    },
    "plugins": ["@typescript-eslint"],
    "rules": {
        "@typescript-eslint/strict-boolean-expressions": ["error"]
    }
}
```

```TS
function f(arg: "Option1" | "Option2" | null) {
    if (arg)
        console.log(`Input: ${arg}`);
    else
        console.log("Null was input");
}
```

tsconfig: the default as output by `tsc --init`

**Expected Result**

I expected there to be no error, as there is exactly one possible falsey value, no ambiguity.

**Actual Result**
```
  2:9  error  Unexpected nullable string value in conditional. Please handle the nullish/empty cases explicitly  @typescript-eslint/strict-boolean-expressions
```
**Additional Info**
This is similar to the likely non-bug #2894, but in that example there are multiple falsey values.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.15.2` |
| `@typescript-eslint/parser`        | `4.15.2` |
| `TypeScript`                       | `4.1.5` |
| `ESLint`                           | `7.20.0` |
| `node`                             | `10.24.0` |

--- a/packages/eslint-plugin/src/rules/no-unnecessary-type-assertion.ts
+++ b/packages/eslint-plugin/src/rules/no-unnecessary-type-assertion.ts
@@ -151,6 +151,10 @@ export default util.createRule<Options, MessageIds>({
               },
             });
           }
+          // for all other = assignments we ignore non-null checks
+          // this is because non-null assertions can change the type-flow of the code
+          // so whilst they might be unnecessary for the assignment - they are necessary
+          // for following code
           return;
         }
 

--- a/packages/eslint-plugin/src/rules/no-unnecessary-type-assertion.ts
+++ b/packages/eslint-plugin/src/rules/no-unnecessary-type-assertion.ts
@@ -137,19 +137,20 @@ export default util.createRule<Options, MessageIds>({
       TSNonNullExpression(node): void {
         if (
           node.parent?.type === AST_NODE_TYPES.AssignmentExpression &&
-          node.parent?.operator === '=' &&
-          node.parent.left === node
+          node.parent.operator === '='
         ) {
-          context.report({
-            node,
-            messageId: 'contextuallyUnnecessary',
-            fix(fixer) {
-              return fixer.removeRange([
-                node.expression.range[1],
-                node.range[1],
-              ]);
-            },
-          });
+          if (node.parent.left === node) {
+            context.report({
+              node,
+              messageId: 'contextuallyUnnecessary',
+              fix(fixer) {
+                return fixer.removeRange([
+                  node.expression.range[1],
+                  node.range[1],
+                ]);
+              },
+            });
+          }
           return;
         }
 

--- a/packages/eslint-plugin/tests/rules/no-unnecessary-type-assertion.test.ts
+++ b/packages/eslint-plugin/tests/rules/no-unnecessary-type-assertion.test.ts
@@ -202,6 +202,12 @@ let a: { b?: string } | undefined;
 a!.b = '';
       `,
     },
+    `
+let value: number | undefined;
+let values: number[] = [];
+
+value = values.pop()!;
+    `,
   ],
 
   invalid: [
@@ -477,14 +483,10 @@ y! = 0;
       output: `
 let x: number | undefined;
 let y: number | undefined;
-y = x;
+y = x!;
 y = 0;
       `,
       errors: [
-        {
-          messageId: 'contextuallyUnnecessary',
-          line: 4,
-        },
         {
           messageId: 'contextuallyUnnecessary',
           line: 5,

